core/trie/bitarray.go

```diff
@@ -0,0 +1,458 @@
+//	a.EqualMSBs(b) = true  // First 4 MSBs match
+//
+//	a = 1100 (len=4)
+//	b = 1101 (len=4)
+//	a.EqualMSBs(b) = false // All bits compared, not equal
+//
+//	a = 1100 (len=4)
+//	b = [] (len=0)
+//	a.EqualMSBs(b) = true  // Zero length is always a prefix match
+func (b *BitArray) EqualMSBs(x *BitArray) bool {
+	if b.len == x.len {
+		return b.Equal(x)
+	}
+
+	if b.len == 0 || x.len == 0 {
+		return true
+	}
+
+	// Compare only the first min(b.len, x.len) bits
+	minLen := b.len
+	if x.len < minLen {
+		minLen = x.len
+	}
+
+	return new(BitArray).MSBs(b, minLen).Equal(new(BitArray).MSBs(x, minLen))
+}
+
+// Sets b to the most significant 'n' bits of x.
+// If n >= x.len, b is an exact copy of x.
+// For example:
+//
+//	x = 11001011 (len=8)
+//	MSBs(x, 4) = 1100 (len=4)
+//	MSBs(x, 10) = 11001011 (len=8, original x)
+//	MSBs(x, 0) = 0 (len=0)
+func (b *BitArray) MSBs(x *BitArray, n uint8) *BitArray {
+	if n >= x.len {
+		return b.Set(x)
+	}
+
+	return b.Rsh(x, x.len-n)
+}
+
+// Sets b to the longest sequence of matching most significant bits between two bit arrays.
+// For example:
+//
+//	x = 1101 0111 (len=8)
+//	y = 1101 0000 (len=8)
+//	CommonMSBs(x,y) = 1101 (len=4)
+func (b *BitArray) CommonMSBs(x, y *BitArray) *BitArray {
+	if x.len == 0 || y.len == 0 {
+		return emptyBitArray
+	}
+
+	long, short := x, y
+	if x.len < y.len {
+		long, short = y, x
+	}
+
+	// Align arrays by right-shifting longer array and then XOR to find differences
+	// Example:
+	//   short = 1100 (len=4)
+	//   long  = 1101 0111 (len=8)
+	//
+	// Step 1: Right shift longer array by 4
+	//   short = 1100
+	//   long  = 1101
+	//
+	// Step 2: XOR shows difference at last bit
+	//   1100 (short)
+	//   1101 (aligned long)
+	//   ---- XOR
+	//   0001 (difference at last position)
+	// We can then use the position of the first set bit and right-shift to get the common MSBs
+	diff := long.len - short.len
+	b.Rsh(long, diff).Xor(b, short)
+	divergentBit := findFirstSetBit(b)
+
+	return b.Rsh(short, divergentBit)
+}
+
+// Sets b = x >> n and returns b.
+//
+//nolint:mnd
+func (b *BitArray) Rsh(x *BitArray, n uint8) *BitArray {
+	if x.len == 0 {
+		return b.Set(x)
+	}
+
+	if n >= x.len {
+		return b.clear()
+	}
+
+	switch {
+	case n == 0:
+		return b.Set(x)
+	case n >= 192:
+		b.rsh192(x)
+		b.len = x.len - n
+		n -= 192
+		b.words[0] >>= n
+	case n >= 128:
+		b.rsh128(x)
+		b.len = x.len - n
+		n -= 128
+		b.words[0] = (b.words[0] >> n) | (b.words[1] << (64 - n))
+		b.words[1] >>= n
+	case n >= 64:
+		b.rsh64(x)
+		b.len = x.len - n
+		n -= 64
+		b.words[0] = (b.words[0] >> n) | (b.words[1] << (64 - n))
+		b.words[1] = (b.words[1] >> n) | (b.words[2] << (64 - n))
+		b.words[2] >>= n
+	default:
+		b.Set(x)
+		b.len -= n
+		b.words[0] = (b.words[0] >> n) | (b.words[1] << (64 - n))
+		b.words[1] = (b.words[1] >> n) | (b.words[2] << (64 - n))
+		b.words[2] = (b.words[2] >> n) | (b.words[3] << (64 - n))
+		b.words[3] >>= n
+	}
+
+	return b
+}
+
+// Sets b = x ^ y and returns b.
```

Comment:
It happens all along the comments. Just minor nitpicks with the comment, if you agree with the suggestion, could you update the remaining ones?